### ISSUE TITLE
Warn on redundant null suppressions

### ISSUE THREAD

**Issue by alrz** [dotnet/csharplang#2884]
There are two scenarios that result in unintended redundant null suppressions: either (1) code is changed and a null suppression is no longer needed, or (2) compiler doesn't produce a false positive anymore. In both cases, the presence of such suppression operator is misleading and should be flagged.

**Comment by DavidArno** [comment]
Oh yes. I sooo want this. Currently, I have to manually work my way through a code base, removing `!`'s one at a time to see if a warning appears. This isn't a good workflow.

Of course, the issue will be that introducing this warning is a breaking change. The team doesn't do breaking changes... :cry:

**Comment by alrz** [comment]
@DavidArno I think NRT improvements will be built on top of warning waves, this would be another case in that bucket.

**Comment by canton7** [comment]
Alternatively, I would be happy with this being a message, rather than a warning.

**Comment by DavidArno** [comment]
@alrz, makes sense.

@canton7, that is another option. I'd prefer a warning, but a message would be far better than the current silence.

:+1:

**Comment by Joe4evr** [comment]
Yeah, I'd say this falls into the same bucket as the Message for redundant parentheses.

**Comment by MarkusAmshove** [comment]
You could change the severity of the message to a warning or an error in `.editorconfig`, so I'm fine with that :-)

**Comment by sharwell** [comment]
My plan here is have the compiler report a hidden diagnostic, similar to the one reported for unused `using` directives. An analyzer could then report that diagnostic with a different ID, similar to IDE0005.

**Comment by DavidArno** [comment]
How does the analyser gain access to that hidden diagnostic? It sounds like a good plan, but I'm confused how it would work.

**Comment by sharwell** [comment]
@DavidArno It's already the strategy we use for IDE0005. See https://github.com/dotnet/roslyn/pull/29704 for the most recent work on that.

📝 I believe this issue can be closed in favor of dotnet/roslyn#34714.

**Comment by DavidArno** [comment]
Thanks for the link, @sharwell. I'll take a read. Great to hear this is already being worked on. 👍

**Comment by alrz** [comment]
>The compiler diagnostic would be reported for all unnecessary suppression operators, while the analyzer would avoid reporting a diagnostic in generated code. 

sounds great! thanks.

**Comment by sharwell** [comment]
> Great to hear this is already being worked on. 👍

It's not already being worked on, but we do have a model for implementation that we know works so it should be unblocked for someone to start working on it. 😄

**Comment by YairHalberstadt** [comment]
> so it should be unblocked for someone to start working on it.

*hint hint*?

**Comment by sharwell** [comment]
😆 I was totally thinking "You know who would probably do this really well? I bet @alrz would."
